pkg/parser/io_parser.go

```diff
@@ -0,0 +1,152 @@
+)
+
+const (
+	newline    = '\n'
+	newlineLen = 1
+)
+
+type IOParser struct {
+	inputBuffer  []byte
+	inputLock    sync.Mutex
+	outputBuffer []byte
+	outputLock   sync.Mutex
+	parse        *Parser
+}
+
+func NewIOParser(bufSize uint64, parserBufSize uint64, debug bool) *IOParser {
+	return &IOParser{
+		inputBuffer:  make([]byte, 0, bufSize),
+		outputBuffer: make([]byte, 0, bufSize),
+		parse:        NewBufferedParser(parserBufSize, debug),
+	}
+}
+
+func (p *IOParser) Read(b []byte) (int, error) {
+	p.outputLock.Lock()
+	defer p.outputLock.Unlock()
+
+	var (
+		bufcap    = cap(b)
+		outbuflen = len(p.outputBuffer)
+	)
+
+	// If somehow we have exactly cap(b) in output buffer, just copy and return.
+	if outbuflen == bufcap {
+		copy(b, p.outputBuffer)
+		p.outputBuffer = p.outputBuffer[:0]
+		return bufcap, nil
+	}
+
+	// If we have more than cap(b), copy, resize and return.
+	if outbuflen > bufcap {
+		copy(b, p.outputBuffer[:bufcap])
+
+		// Proper way to resize the slice without losing it's length
+		copy(p.outputBuffer, p.outputBuffer[bufcap:])
+		p.outputBuffer = p.outputBuffer[:outbuflen-bufcap]
+
+		return bufcap, nil
+	}
+
+	// Not enough data left in output buffer, so drain it unconditionally.
+	copy(b[:outbuflen], p.outputBuffer)
+	p.outputBuffer = p.outputBuffer[:0]
+
+	// Remember how much we can write more
+	written := outbuflen
+
+	// Loop until we have either exhausted the parser OR the buffer.
+	for {
+		res, more := <-p.parse.Results()
+		if !more {
+			return written, io.EOF
+		}
+
+		data, err := json.Marshal(res)
+		if err != nil {
+			return written, err
+		}
+
+		var (
+			datalen = len(data)
+			capleft = bufcap - written
+		)
+
+		// We have exactly enough data to fill in the rest of the buffer, so copy and return.
+		if datalen == capleft {
+			copy(b[written:], data)
+			return bufcap, nil
+		}
+
+		// We have more data to write than we can fit in the output.
+		// Copy however many bytes we can from serialized result and store the rest in output buffer.
+		if datalen > capleft {
+			copy(b[written:], data[:capleft])
+			p.outputBuffer = append(p.outputBuffer, data[capleft:]...)
+			return bufcap, nil
+		}
+
+		// We can fit more than we have available for writing.
+		// Write as much as we can and continue to the next iteration.
+		copy(b[written:], data)
+		written = written + datalen
+	}
+}
+
+func (p *IOParser) Write(b []byte) (int, error) {
+	p.inputLock.Lock()
+	defer p.inputLock.Unlock()
+
+	written := 0
+
+	for {
+		buflen := len(b)
+
+		// Quick check for empty buffer.
+		if buflen == 0 {
+			return written, nil
+		}
+
+		// Check if incoming buffer contains a newline,
+		// since it indicates that we can now send the utterance to AST parser.
+		linesep := bytes.IndexRune(b, newline)
+
+		// If incoming bytestring doesn't contain a newline,
+		// then just append it to our "current line" buffer and return.
+		if linesep == -1 {
+			p.inputBuffer = append(p.inputBuffer, b...)
+			return written + buflen, nil
```

Comment:
For clarity this should be just `buflen`

Reply:
No, this will not work, because `buflen` is not the full length of the buffer, but just the slice of `b`, that current iteration of the loop is working with.

Reply:
Ah yeah, I kinda missed that this is in a loop. Which also confused me in the comment below. Both are resolved for me.